`spack install` gracefully handles foreground/background

tgamblin opened the pull request:
Fixes #15620.

This resolves the second issue in #15620 by reworking #14682.

With the addition of #13100, people are going to try to run Spack in the background more often.  This makes Spack handle that situation gracefully, the way a good POSIX program should.

Specifically:
1. When `spack install` is running, we disable echo and canonical input so that users can type `v` to toggle build output.  We do that in a safe way now, so that it does not generate `SIGTTOU` in the background (#14682 did this).

2. We properly disable keyboard input mode when Spack is placed in the background, and re-enable it when Spack is in the foreground.  This means that if you Ctrl-Z a spack install, your terminal won't be left in a weird state.

3. ~~Spack won't write verbose install output in the background.  This is mostly because we already write a log file -- that's what the log is for (if this is controversial, we can re-enable it, but it's fairly convenient to be able to ctrl-Z/bg/fg a running spack install without getting build spew).~~

   We'll continue writing verbose output when Spack is in the background. If you have `stty +tostop` on, it'll end up stopping the build when you try to run in the background, unless you redirect output. This is normal POSIX behavior, and it allows you to `spack install -v &> log.txt &`, which our users do.  You may want this for parallel build output as well.

4. Spack works fine when stopped in the background or when running in the background.

(2) is handled mostly with signal handlers (the way things like `vi` and `emacs` do it) -- see the code for how that's done -- it's a bit tricky in Python, as Python did not support blocking signals with masks until 3.8.  It turns out we can still make it work.

This means you can do stuff like this (use `libiconv` b/c the build is long but not too long, and `--overwrite` to test it multiple times):
```console
spack install --overwrite -y libiconv
...
[type v]
checking for /Users/gamblin2/src/spack/lib/spack/env/clang/clang option to produce PIC... -fno-common -DPIC
checking if /Users/gamblin2/src/spack/lib/spack/env/clang/clang PIC flag -fno-common -DPIC works... yes
checking if /Users/gamblin2/src/spack/lib/spack/env/clang/clang static flag -static works... no
[Ctrl-Z]

[1]+  Stopped                 spack install --overwrite -y libiconv
$ bg
[1]+ spack install --overwrite -y libiconv &
checking for archiver @FILE support... no
checking for strip... strip
$ fg
spack install --overwrite -y libiconv
checking for ranlib... ranlib
checking for gawk... no
...
==> 29825: libiconv: Successfully installed libiconv
  Fetch: 0.02s.  Build: 1m 10.03s.  Total: 1m 10.04s.
[+] /Users/gamblin2/src/spack/opt/spack/darwin-mojave-x86_64/clang-11.0.0-apple/libiconv-1.16-jg6ekuhbklculym7pyjpx6l5jetym7wp
```

You should be able to do any combination of transitions and have output work as you expect, e.g.:

1. start in foreground, ctrl-z, fg, v, ctrl-z, fg, v, v, v, ctrl-z, bg, fg, v
2. start in background, fg, v, ctrl-z, fg, v, v, v, ctrl-z, bg, fg, v

I think that covers all the transitions.  Maybe @adamjstewart can try hard to break it.

adamjstewart commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,126 @@ def __init__(self, stream):
         self.stream = stream
 
+    def _is_background(self):
+        """True iff """
```
Is this missing more text?

tgamblin replied:
fixed

adamjstewart commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,126 @@ def __init__(self, stream):
+
+    def _get_canon_echo_flags(self):
+        """Get current termios canonical and echo settings."""
+        cfg = termios.tcgetattr(self.stream)
```
I notice at the top of the file we have a try-except for importing `termios`. Do we need to check whether or not this is None first?

tgamblin replied:
this is only called from one place, if `self.old_cfg` is set, and that is only set when we have `termios`

tgamblin commented:
Update: I changed (3) to be a) simpler and b) more unixy.  I also think that if we keep (3) the way we originally had it, it'll interfere with the way a some users log their spack builds (@cyrush @white238).  The new semantics are more POSIX-y and should make more sense.

@becker33: FYI

tgamblin commented:
> While you're digging around in here... how hard would #12206 be to implement?

It's not hard to let you toggle that here, but:
1. That `v` is implemented in the logger process, and it's only active while the build is going on.  You really want to be able to type `d` whenever `spack` is running -- there are way more `tty.debug()` statements outside the build than in it.
2. Because it's in the logger process, it can't affect the parent Spack without a major rework -- the `tty.debug()` calls are all in the parent but we can only change the var in the child.

adamjstewart commented:
One of the linux builds is failing with an interesting error message:
```
==> 8297: lz4: Building lz4 [MakefilePackage]
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/home/runner/work/spack/spack/lib/spack/llnl/util/tty/log.py", line 578, in _writer_daemon
    kb.check()
AttributeError: 'NoneType' object has no attribute 'check'
==> Error: ProcessError: Command exited with status -13:
    'make' '-j2' 'MOREFLAGS=-lrt'
See build log for details:
  /tmp/runner/spack-stage/spack-stage-lz4-1.9.2-6ncwqjtyphpdvi36jyhrg3bwvipzstvi/spack-build-out.txt

==> Error: Failed to install lz4 due to ChildError: ProcessError: Command exited with status -13:
    'make' '-j2' 'MOREFLAGS=-lrt'See build log for details:
Traceback (most recent call last):
  File "/home/runner/work/spack/spack/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/home/runner/work/spack/spack/lib/spack/spack/installer.py", line 1113, in build_process
    phase(pkg.spec, pkg.prefix)
  File "/home/runner/work/spack/spack/lib/spack/spack/package.py", line 108, in phase_wrapper
    phase(spec, prefix)
  File "/home/runner/work/spack/spack/var/spack/repos/builtin/packages/lz4/package.py", line 39, in build
    else:
  File "/home/runner/work/spack/spack/lib/spack/spack/build_environment.py", line 131, in __call__
    return super(MakeExecutable, self).__call__(*args, **kwargs)
  File "/home/runner/work/spack/spack/lib/spack/spack/util/executable.py", line 188, in __call__
    raise ProcessError('Command exited with status %d:' %
spack.util.executable.ProcessError: Command exited with status -13:
    'make' '-j2' 'MOREFLAGS=-lrt'

##[error]Process completed with exit code 1.
```

tgamblin commented:
Oh that’s a refactoring bug that crept in. When I tried to make the naming and docs better. Ugh. I’ll fix that when I get home. 😬

tgamblin commented:
Ok wasn't actually a refactoring error -- but fixed anyway.  Try now.

tgamblin commented:
I'm trying to add some tests now for the signal handling part.

scheibelp commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
-                termios.tcsetattr(self.stream, termios.TCSADRAIN, self.new_cfg)
-                self.old_cfg = old_cfg
+            # add handlers to disable/enable keyboard input when process
+            # moves from foreground to background.
```
I'd replace "from foreground to background" with "between foreground and background"

tgamblin replied:
done

scheibelp commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
+        os.kill(os.getpid(), signal.SIGTSTP)
+
+    def _cont_handler(self, signum, frame):
+        """Handler to restore or enable term settings on SIGCONT."""
```
The docstring diagram for this class is excellent, I think it would be great to refer to it here.

tgamblin replied:
This handler no longer exists -- there's just one.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -38,32 +39,40 @@
-    if sys.stdout.isatty():
-        return os.getpgrp() != os.tcgetpgrp(sys.stdout.fileno())
-    return False  # not writing to tty, not background
+def _is_background_tty(stream):
```
> This is not correct because of the check for stream.isatty()

I'm not sure how it's not correct.  `os.getpgrp()` gets my process's group id.  It's checking whether or not it's in the foreground for the terminal attached to `stream`.

scheibelp replied:
When you say "IFF" do you mean "if and only if"?

tgamblin replied:
yes

scheibelp replied:
Ah so what I mean when I say this is wrong is that this function can return `False` if `stream.isatty() == False`, but it is possible that there are non-tty FG streams.

So I think the following wording is better:

> Return True if the stream is a tty and is in the background.

In that wording, you could also replace "if" with "iff", but I think it may just be distracting.

tgamblin replied:
I kind of don't like that wording because "the stream" is never in the background.  It's *the calling process* that is either fg or bg.  Do you have a suggestion for how to work that in?

scheibelp replied:
> Return True if the stream is a tty and the calling process is in the background

tgamblin replied:
done.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
+        try:
+            termios.tcsetattr(self.stream, termios.TCSANOW, new_cfg)
+        except termios.error as e:
+            # TODO: does this happen? It was needed at one point, but I
```
The prior code (below, red line 120) said:

> # some OS's do not support termios, so ignore

But the code isn't actually called unless we have `termios`.  I am not sure we really *need* the `except` block anymore and am somewhat inclined to delete it, as I haven't actually seen it get triggered.  I think I will do that instead.

tgamblin replied:
I just deleted this, and it causes no issues in the tests.  I think it's a remnant of some prior misuse of `termios`

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -445,62 +571,47 @@ def _writer_daemon(self, stdin):
                     # No need to set any timeout for select.select
                     # Wait until a key press or an event on in_pipe.
                     rlist, _, _ = select.select(istreams, [], [])
+
```
This call is no longer necessary.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -445,62 +571,47 @@ def _writer_daemon(self, stdin):
                             break
 
+                        # find control characters and strip them.
+                        controls = control.findall(line)
```
Because it's far simpler this way.  `handle_write` was extremely confusing.

1. It did not just handle a write -- It read the line and returned something based on the result that would trigger a break (or not) in the surrounding loop.  That made the control flow hard to follow.
2. It shadowed `force_echo` and returned a value that would modify `force_echo` in the surrounding scope.
3. It doesn't really stand on its own -- you *have* to understand the loop below to understand what it does.

For those reasons I think it just confuses the control flow here.  Add to that that the writer daemon is persistent and manages long-lived state (like `echo` and `force_echo`) as local variables, and I think there is already enough control complexity with the daemon function and the loop.  It makes more sense to have simpler control flow here than separation of concerns/encapsulation.

If anything it might make sense to make the daemon a class instead of a function (so the state management is clearer) -- but I think that is a larger refactor.

scheibelp replied:
This is OK (generally I prefer while loops with substantial logic refactor into a while loop with a function call, but the logic here doesn't quite make the threshold of "too much"), but overall these arguments feel like they are targeting symptoms rather than the cause. I think most of these complaints could be resolved by making a separate function with a clear name and an actual docstring.

> It did not just handle a write -- It read the line and returned something based on the result that would trigger a break (or not) in the surrounding loop

That's if the stream ends. This is a common theme of while loops that read streams. I don't think it's overly complex to put in a function that reads a stream and can return the possibility that the stream is finished.

> It shadowed force_echo and returned a value that would modify force_echo in the surrounding scope.

That's because it was defined inside the function rather than as a separate function.

> It doesn't really stand on its own -- you have to understand the loop below to understand what it does.

docstring

tgamblin replied:
I'm still not convinced and am not interested in re-introducing what I think is an overly complex function at this point.

tgamblin replied:
yeah ok -- I think it makes sense to make this a separate function.

tgamblin replied:
This is done.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
+        with ignore_signal(signal.SIGTTOU):
+            termios.tcsetattr(self.stream, termios.TCSANOW, self.old_cfg)
+
+    def check_fg(self):
```
This is gone.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
+
+    def _restore_input(self):
+        """Restore the original input configuration on ``self.stream``."""
+        assert self.old_cfg, "_restore_input only works with termios"
```
I just removed these assertions -- I'd really need to add them to *more* functions if I wanted to be complete, and as they are they just confuse things.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
-                self.old_cfg = old_cfg
+            # add handlers to disable/enable keyboard input when process
+            # moves from foreground to background.
+            signal.signal(signal.SIGTSTP, self._tstp_handler)
```
I've simplified the signal registration and handlers quite a bit in the latest pushes.

tgamblin commented:
@scheibelp @becker33 @adamjstewart: this is ready for a re-review.  All the testing is there, and the signal handlers have been reworked.

scheibelp commented on lib/spack/external/defer_signals.py:
```diff
@@ -0,0 +1,97 @@
+    Takes optional list of signals to defer (default: SIGHUP, SIGINT, SIGTERM).
+    Signals can be identified by number or by name.
+
+    Allows you to wrap instruction sequences that ought to be atomic and ensure
```
I don't think this is possible: the OS can context switch without signals.

tgamblin replied:
I agree but this is an external library, and we do not actually care if we context-switch here -- we only care about the main thread in the daemon process -- the one that receives signals.

scheibelp replied:
I see what you mean and that's pretty cool, but to make sure, does is the signal *required* to actually change the state of the process? For example, where this is used, there is some code:

```
        with defer_signals(...):
             bg = self._is_background()
             # <--- say we switch from background to foreground here
             if not bg and any(flags):    # fg, but input not enabled
```

My understanding is that the process can switch between the BG and FG in between collecting the state and checking the state. The signal might not be received (or rather the handler might be delayed), but the state of the process has changed.

So I think this logic may be necessary but I'm not sure if it's sufficient.

scheibelp commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -88,43 +129,122 @@ def __init__(self, stream):
-                self.old_cfg = old_cfg
+            # this itimer handles cases where the OS doesn't notify us of
+            # a background or foreground change.
+            signal.setitimer(signal.ITIMER_REAL, 1e-3, 1e-3)
```
I think this can be `1e-1`: one tenth of a second is sufficiently responsive for humans.

tgamblin replied:
I could do 1e-2, but I think 1e-1 is too coarse -- humans can hit keys faster than that.

tgamblin replied:
ok looked at this in more details -- 1e-1 is probably a good idea.  1e-3 makes the daemon use ~16% CPU ; 1e-2 is 2-4% and 1e-1 is ~0.5% CPU.  And I can't tell a difference with `v`.

tgamblin replied:
updated to 1e-1

scheibelp commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -88,43 +129,122 @@ def __init__(self, stream):
+            if not bg and any(flags):    # fg, but input not enabled
+                self._enable_keyboard_input()
+            elif bg and not all(flags):  # bg, but input enabled
+                self._restore_input()
```
This function name (`_restore_input`) keeps confusing me and I have to look up what it does. I think `_restore_default_terminal_settings` would be better.

tgamblin replied:
done

scheibelp commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -88,43 +129,122 @@ def __init__(self, stream):
+            bg = self._is_background()
+
+            # restore sanity if any flags are amiss
+            if not bg and any(flags):    # fg, but input not enabled
```
Can you add comments that refer to the excellent state diagram above (e.g. this corresponds to `SIGALRM` handling)?

tgamblin replied:
It's a bit more complicated now that everything uses the same handler -- the same calls are now used for SIGTSTP and SIGCONT -- so this isn't just for SIGALRM.

I added "see diagram in class docs"

scheibelp replied:
I think in general everything may not map 1-to-1, although my understanding was that this particular block (inside the if) was for handling SIGALRM: based on the diagram, I wouldn't expect to be here otherwise.

tgamblin replied:
This is outdated -- no more SIGALRM

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -445,62 +565,58 @@ def _writer_daemon(self, stdin):
+                                if stdin.read(1) == 'v':
+                                    echo = not echo
+                            except IOError as e:
+                                # if SIGTTIN is ignored,
```
yeah I forgot to finish this comment. The protocol is that a read in the bg will generate `SIGTTIN`, and if you ignore `SIGTTIN`, the call returns 1 with `EIO` to let you know it failed (instead of the signal).  We

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -445,62 +565,58 @@ def _writer_daemon(self, stdin):
+                    try:
+                        rlist, _, _ = select.select(istreams, [], [])
+                    except select.error as e:
+                        # This happens we get a signal while in select()
```
done

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -88,43 +129,122 @@ def __init__(self, stream):
+
+    @contextmanager
+    def defer_signals(self):
+        with defer_signals(self.signals):
```
there's no issue here.  one is `self.defer_signals` and the other is `defer_signals`

scheibelp replied:
I'm aware that it is functionally correct: I'm saying it's problematic because it's hard to read.

tgamblin replied:
This is gone in the latest revision.

tgamblin commented:
@adamjstewart: can you verify that this still works on cray?

scheibelp commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+    """
+    os.setsid()   # new session; this process is the controller
+
+    master_fd, child_fd = os.openpty()
```
Note that https://docs.python.org/3/library/os.html (specifically the section on `openpty`) mentions that these are non-inheritable, which is a problem on Windows (and probably also Python >= 3.8 on Mac OS). The doc suggests that the `pty `module may be more portable, but I don't know if that is useful here.

tgamblin replied:
They're not actually passed to the child process, so I think this is safe.

scheibelp commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+    os.close(pty_fd)
+
+    ready = multiprocessing.Value('i', False)
+    proc = multiprocessing.Process(
```
I think `child_process` would be a better name for keeping track of which process is which (and for that matter `_child_process` could be called `_set_up_and_run_child_function`)

tgamblin replied:
Done.

adamjstewart commented:
@tgamblin I'm seeing a problem when I try to install something from the command line. I'm not even pressing <kbd>v</kbd>.
```console
$ spack install libsigsegv
==> 45064: Installing libsigsegv
==> Using cached archive: /mnt/a/u/sciteam/stewart1/spack/var/spack/cache/_source-cache/archive/3a/3ae1af359eebaa4ffc5896a1aee3568c052c99879316a1ab57f8fe1789c390b6.tar.gz
==> Staging archive: /scratch/sciteam/stewart1/spack-stage-libsigsegv-2.12-4otz7mr7kc7sxlehzv4szo7zpnudfqth/libsigsegv-2.12.tar.gz
==> Created stage in /scratch/sciteam/stewart1/spack-stage-libsigsegv-2.12-4otz7mr7kc7sxlehzv4szo7zpnudfqth
==> No patches needed for libsigsegv
==> 45064: libsigsegv: Building libsigsegv [AutotoolsPackage]
==> 45064: libsigsegv: Executing phase: 'autoreconf'
==> 45064: libsigsegv: Executing phase: 'configure'
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/llnl/util/tty/log.py", line 598, in _writer_daemon
    line = in_pipe.readline()
IOError: [Errno 4] Interrupted system call
```
At this point, the code hangs.

scheibelp commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+        self.wait(lambda: "T" not in status)
+
+
+def _child_process(
```
While it is possible for `_child_process` and `_master_process` to stand alone completely, IMO it would be easier to read if they were static methods of `PseudoShell`.

tgamblin replied:
Done

scheibelp commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+        Args:
+            pid (int): id of process to control
+            master_fd (int): master file descriptor created as with
+                ``os.openpty()``, attached to pid's stdin
```
Ideally the fact that the fd was created with `openpty` would not need to be understood by this object: IMO that detail should be omitted here.

tgamblin replied:
fixed

scheibelp commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,472 @@
+    ctl.wait_disabled()
+
+
+def _mock_shell_integration_no_termios(proc, ctl, attrs):
```
This should mention that it doesn't use `proc` or `attrs` (in order to conform to an API). Or throw a boilerplate comment on this and all associated functions. Or put them in a class and document at the class level.

tgamblin replied:
Done

scheibelp commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+            sys.stderr.write("\n")
+
+    def input_on(self):
+        """True if the temrchild process """
```
"temrchild"?

tgamblin replied:
fixed

scheibelp commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+
+    proc.join()
+
+    return error or proc.exitcode
```
Can you update the docstring to mention what is being returned? I think the idea is that you are returning whether the parent or the child process failed.

tgamblin replied:
done

tgamblin commented:
@adamjstewart: try the latest push -- I [learned a thing](https://www.python.org/dev/peps/pep-0475/#interrupted-system-calls) 😱.

adamjstewart commented:
@tgamblin new error msg:
```console
$ spack install libsigsegv
==> 20475: Installing libsigsegv
==> Using cached archive: /mnt/a/u/sciteam/stewart1/spack/var/spack/cache/_source-cache/archive/3a/3ae1af359eebaa4ffc5896a1aee3568c052c99879316a1ab57f8fe1789c390b6.tar.gz
==> Staging archive: /scratch/sciteam/stewart1/spack-stage-libsigsegv-2.12-4otz7mr7kc7sxlehzv4szo7zpnudfqth/libsigsegv-2.12.tar.gz
==> Created stage in /scratch/sciteam/stewart1/spack-stage-libsigsegv-2.12-4otz7mr7kc7sxlehzv4szo7zpnudfqth
==> No patches needed for libsigsegv
==> 20475: libsigsegv: Building libsigsegv [AutotoolsPackage]
==> 20475: libsigsegv: Executing phase: 'autoreconf'
==> 20475: libsigsegv: Executing phase: 'configure'
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/llnl/util/tty/log.py", line 584, in _writer_daemon
    rlist, _, _ = select.select(istreams, [], [])
error: (4, 'Interrupted system call')
```

tgamblin commented:
@adamjstewart: which python version are you using?  Is this on blue waters?

tgamblin commented:
@adamjstewart: I added *back* the guard around `select()` that I had removed earlier, as it seems that even if we call `signal.siginterrupt(signum, False)`, we still end up getting `EINTR`.  This is kind of strange to me, but see if this change fixes your first issue.  If not, I'll need to look at that.

adamjstewart commented:
@tgamblin new error msg:
```console
$ spack install libsigsegv
==> 59394: Installing libsigsegv
==> Using cached archive: /mnt/a/u/sciteam/stewart1/spack/var/spack/cache/_source-cache/archive/3a/3ae1af359eebaa4ffc5896a1aee3568c052c99879316a1ab57f8fe1789c390b6.tar.gz
==> Staging archive: /scratch/sciteam/stewart1/spack-stage-libsigsegv-2.12-4otz7mr7kc7sxlehzv4szo7zpnudfqth/libsigsegv-2.12.tar.gz
==> Created stage in /scratch/sciteam/stewart1/spack-stage-libsigsegv-2.12-4otz7mr7kc7sxlehzv4szo7zpnudfqth
==> No patches needed for libsigsegv
==> 59394: libsigsegv: Building libsigsegv [AutotoolsPackage]
==> 59394: libsigsegv: Executing phase: 'autoreconf'
==> 59394: libsigsegv: Executing phase: 'configure'
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/mnt/a/u/sciteam/stewart1/spack/lib/spack/llnl/util/tty/log.py", line 612, in _writer_daemon
    line = in_pipe.readline()
IOError: [Errno 4] Interrupted system call
```
Yes, this is on BW, with Python 2.6.9. The problem does not occur if I use a Spack-installed Python 3.7.6.

scheibelp commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,472 @@
+    split = re.split(r'\s+', stripped) if stripped else []
+
+    numbers = set([int(n) for n in split])
+    all_numbers = set(range(shell.attrs["count"]))
```
`expected_numbers`? IMO "expected" is common in variable names that are used to store the state we think ought to be.

tgamblin replied:
fixed

scheibelp commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,472 @@
+    assert 0 in numbers
+
+    # all numbers are really output we expect
+    assert numbers <= all_numbers
```
I think I see that this test operates by constantly outputting new numbers while switching terminal output, but this approach feels a bit random: would it be possible to use interprocess signals to have the child process note when it reads a specific number, and we could ensure that exactly the numbers we expect are present?

tgamblin replied:
This is hard to set up because `v` is fundamentally asynchronous w.r.t. the output of the logged process.  I think the test as-is is pretty simple.  I do not see a way to make it more specific as you suggest without making it very invasive.

scheibelp replied:
As it is, I think there are many possible ways this test can fail without it being obvious in the test output, so my goal is to figure out how to make this more precise.

I agree my initial suggestion is lacking: I've had trouble in the past when I tried coordinating with interprocess signals to check output. The problem is that it is hard to reason about whether the signals are interfering or creating an artificial result.

I think it might be easier to improve this test with locking: we can create a lock that is held by the parent process whenever it is in "v" mode. The child process checks this lock and if it is held, then it outputs "2", and if it is not, then it outputs "1". In more detail:

* There would actually have to be two locks, one for when "v" is applied and one for when the child process is going to write something, let's call that "w"
* The parent process must hold "w" before it acquires or releases "v"
* The child process must hold "w" before it writes anything

Does that seem more reasonable?
I think that would be easier to reason about than IPC signals.

tgamblin replied:
Can you elaborate on the non-obvious ways this could fail?  Essentially this types v, sleeps, and types v again, and it looks to see that something was output. It actually needs a check to make sure more than just 0 was output, but I am not seeing how your approach helps. It would require me to modify log_output, at which point I’m not really testing the “real” code anymore. I don’t think it’s worth it.  I also don’t see why it needs to be “precise”. The feature is not precise. We are testing whether there is output or not, not what that output is, and that is fine here.

scheibelp replied:
I'm not talking about how this test can fail: I'm saying that there are many things that could go wrong in the source itself, and this test would not catch it, or it would fail and it would be difficult to understand what went wrong.

> It would require me to modify log_output

I don't think either of my suggested approaches require that. I'm talking about coordinating between the entity that sends the "v" signal, and the child process that is doing the printing.

> I also don’t see why it needs to be “precise”

If anything goes wrong, and the test is not precise, then it is difficult to figure out how to fix it: anyone editing this code or fixing bugs in the future will have trouble understanding how to fix the test.

> We are testing whether there is output or not, not what that output is

There are many ways we can end up with no output. Most of them are bugs. We want to end up with no output in exactly the way we desire. Being precise about what is captured will help us understand whether we actually succeeded.

tgamblin replied:
@scheibelp: the tests have been simplified and use locks as you suggest, but I do not think this is so much better than before.  In particular, locks don't get rid of the races here.  It is still possible to have a race where the child process notices `v` on `master_fd` *after* the lock takes effect.

I implemented this so that the child writes `off` when `v` is not enabled and `on` when it is -- you can still get some `off`'s in the output, and I do not see a non-invasive way around that, nor do I really think we should try hard to do that.

The new version does make no output unambiguously wrong.  So that's good.  But I had to put a kludgy condition on the assertions -- I allow up to 2 `off`'s in the output.  I think this is about the best we can do and would rather not spend a lot more time on it.  The important piece of this test is that it verifies that `v` toggles output on and off.  I think the current tests do that well enough.

scheibelp commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,472 @@
+    (mock_shell_tstp_tstp_cont, no_termios),
+    (mock_shell_tstp_tstp_cont_cont, no_termios),
+])
+def test_foreground_background_simple(test_fn, termios_on_or_off):
```
By "simple" I think we mean "doesn't fail" - I'd prefer the name be updated. I think it's also worth explaining in a docstring that this is to check for issues like `SIGTOU` raising an error for the running process.

tgamblin replied:
done

tgamblin commented:
So, I've reverted this to use polling.  SIGALRM seemed like a good idea but it fails in very strange ways in Python 2.6 as Adam saw.  The relevant issues are here:

* https://www.python.org/dev/peps/pep-0475/
* https://bugs.python.org/issue8354

It also required additional complexity for `defer_signals`, which we don't really need.

I've reverted this to use polling to handle that last bg -> fg transition, which should fix the Python 2.6 issues.  This does mean that if a build doesn't generate much output, our terminal settings may not be reset for a while (we only check with every line out output) but I don't think that is a big deal.  FWIW, I also tried handling things with a thread to monitor the fg/bg status, and that reduced the total number of signals, but it was hard to make it reliable due to races bt/w the thread and signal handlers.

Polling in the daemon loop seems like the best, most stable, and most compatible way to make this work.

adamjstewart commented:
Linux builds are failing with the following error message:
```
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/home/runner/work/spack/spack/lib/spack/llnl/util/tty/log.py", line 586, in _writer_daemon
    kb.check_fg_bg()
  File "/home/runner/work/spack/spack/lib/spack/llnl/util/tty/log.py", line 213, in check_fg_bg
    flags = self._get_canon_echo_flags()
  File "/home/runner/work/spack/spack/lib/spack/llnl/util/tty/log.py", line 153, in _get_canon_echo_flags
    cfg = termios.tcgetattr(self.stream)
termios.error: (25, 'Inappropriate ioctl for device')
```
Preliminary testing on BW seems fine though.

tgamblin commented:
Linux builds are fixed -- forgot to guard against `not isatty()`

adamjstewart commented:
Now Travis is broken:
```
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/home/travis/build/spack/spack/lib/spack/llnl/util/tty/log.py", line 586, in _writer_daemon
    kb.check_fg_bg()
  File "/home/travis/build/spack/spack/lib/spack/llnl/util/tty/log.py", line 209, in check_fg_bg
    if not termios or not self.stream.isatty():
AttributeError: 'NoneType' object has no attribute 'isatty'
```

tgamblin commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+            bool(cfg[3] & termios.ECHO),
+        )
+
+    def hline(self, name):
```
Done

tgamblin commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+    def wait(self, condition):
+        start = time.time()
+        while (((time.time() - start) < self.timeout) and not condition()):
+            time.sleep(1e-2)
```
Done.

tgamblin commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+            )
+
+    def status(self):
+        """Print debug messgae with status info for the child."""
```
fixed

tgamblin commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,472 @@
+
+
+def test_log_python_output_with_python_stream(capsys, tmpdir):
+    # pytest's DontReadFromInput object does not like what we do here, so
```
I think that's a follow-on PR thing.

scheibelp replied:
This module has been entirely recreated from scratch: IMO now is the best time to consider whether to do it.

tgamblin replied:
I’d rather not do this refactor. I don’t know why we should implement this outside pytest when it’s intended to be there for testing.

scheibelp replied:
> Does capsys.disabled() effectively undo all modifications done by PyTest to sys.stdout, sys.stdin etc.?

Based on looking at the source code of `_pytest/capture` (and specifically the `FDCapture` and `SysCapture` objects, I think the answer to this is "yes".

That leaves the second question. Like I mention above, I think that issue can be resolved by using `capsys.disabled` for all tests in this module.

tgamblin replied:
These are refactored and no longer use `capsys`.

scheibelp replied:
My overall concern is that our fd management and PyTest's fd management (whether it be capsys or capfd) will create an environment that doesn't match where the code actually runs. Overall from reading `FDCapture` I think if it were disabled with `capfd.disabled()` then it wouldn't be an issue. I don't think it's an issue for tests using `PseudoShell` (since it explicitly configures the FD's for the child to inherit).

tgamblin commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,472 @@
+def no_termios():
+    saved = llnl.util.tty.log.termios
+    llnl.util.tty.log.termios = None
+    yield
```
fixed.

tgamblin commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,472 @@
+
+
+@pytest.mark.skipif(not which("ps"), reason="requires ps utility")
+@pytest.mark.skipif(not termios, reason="requires termios support")
```
They use `termios` to test whether the terminal is disabled.  So while the tested code does not use `termios`, the tests need it to validate what happens.

tgamblin commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+
+
+def _master_process(master_function, child_function, attrs):
+    """Master process wrapper for PseudoShell.
```
fixed

tgamblin commented on lib/spack/llnl/util/tty/pty.py:
```diff
@@ -0,0 +1,287 @@
+
+    # call the master function once the child is ready
+    try:
+        controller = ProcessController(
```
Done

scheibelp commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,422 @@
+            print('logged')
+
+        with open('foo.txt') as f:
+            assert f.read() == 'logged\n'
```
(question) why does this not say `'echo\nlogged\n'`?

tgamblin replied:
Yeah good question.  I had just left this in place, as these tests have been present since the original version of this code.  But it looks like this is because of interference like you're describing.  It only seems to happen when using a subprocess (`echo`) so I split this specific test into two -- one to test the logged output (with `capfd.disabled()`) and one to test the echoed output (with `capfd`).  This does not seem to be necessary for the ones that just call `print()`.

tgamblin commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,422 @@
+    with open(log_path) as log:
+        log = log.read().strip().split("\n")
+
+    # output should contain only "on" lines, along with forced output,
```
I merged this with the existing comment.  I think <- 2 offs will be pretty good and is a tighter bound than the test you're proposing -- it's really a function of the `v` latency, not something proportional to the number of ons or offs.  I generally see zero, *maybe* one `off` in the output.

tgamblin commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,422 @@
+
+
+def mock_shell_v_v(proc, ctl, **kwargs):
+    """Child function for test_foreground_background_* below."""
```
good catch -- fixed.

tgamblin commented on lib/spack/spack/test/llnl/util/tty/log.py:
```diff
@@ -0,0 +1,422 @@
+
+    log_path = kwargs["log_path"]
+    write_lock = kwargs["write_lock"]
+    v = kwargs["v"]
```
done

tgamblin commented on lib/spack/spack/test/llnl/util/lang.py:
```diff
@@ -130,3 +130,9 @@ def test_load_modules_from_file(module_path):
+
+
+def test_uniq():
+    assert [1, 2, 3] == llnl.util.lang.uniq([1, 1, 1, 1, 2, 2, 2, 3, 3])
```
done

tgamblin commented:
@scheibelp: I think this addresses everything.

adamjstewart commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -89,42 +150,97 @@ def __init__(self, stream):
+        return self
 
     def __exit__(self, exc_type, exception, traceback):
         """If termios was avaialble, restore old settings."""
```
```suggestion
        """If termios was available, restore old settings."""
```
